File: 57.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
	bool judge(vector<Interval>& intervals, int i) {
		if(i == intervals.size() - 1) return 0;
		if(intervals[i].end >= intervals[i + 1].start) {
			intervals[i].end = intervals[i].end > intervals[i + 1].end ? intervals[i].end : intervals[i + 1].end;
			cout << intervals[i].start << " " << intervals[i].end << endl;
			intervals.erase(intervals.begin() + i + 1);
			return 1;
		}
		return 0;
	}

    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        if(intervals.empty()) {
            intervals.push_back(newInterval);
            return intervals;
        }
      	for(int i = 0; i < intervals.size(); ++i) {
      		if(intervals[i].end < newInterval.start) continue;
      		else if(intervals[i].start > newInterval.end) {
      			intervals.insert(intervals.begin() + i, newInterval);
      			return intervals;
      		}else {
      			intervals[i].start = intervals[i].start < newInterval.start ? intervals[i].start : newInterval.start;
      			intervals[i].end = intervals[i].end > newInterval.end ? intervals[i].end : newInterval.end;
      			cout << intervals[i].start << " " << intervals[i].end << endl;
      			while(judge(intervals, i));
                return intervals;
      		}
      	}
      	intervals.push_back(newInterval);
      	return intervals;  
    }
};

vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
	intervals.push_back(newInterval);
    sort(intervals.begin(), intervals.end(), [](Interval a, Interval b){
    	return a.start < b.start;
    })
    vector<Interval> res;
    res.push_back(intervals[0]);
    for(auto tem : intervals) {
    	int r = res.size() - 1;
    	if(tem.end <= res[r].end) continue;
    	if(tem.start > res[r].end) res.push_back(tem);
    	else if(tem.start <= res[r].end) res[r].end = tem.end;
    }
    return res;
}

File: 57.py
from typing import List


class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if not intervals:
            return [newInterval]
        if newInterval[-1] < intervals[0][0]:
            return [newInterval] + intervals
        if newInterval[-1] == intervals[0][0]:
            intervals[0][0] = newInterval[0]
            return intervals
        if newInterval[0] > intervals[-1][1]:
            return intervals + [newInterval]
        if newInterval[0] == intervals[-1][1]:
            intervals[-1][1] = newInterval[1]
            return intervals
        if newInterval[0] <= intervals[0][0]:
            intervals[0][0] = newInterval[0]
        for i in range(len(intervals)):
            if intervals[i][0] <= newInterval[0] <= intervals[i][1]:
                left = i
                break
            if i < len(intervals) - 1 and intervals[i][1] < newInterval[0] < intervals[i+1][0]:
                if newInterval[1] >= intervals[i+1][0]:
                    intervals[i+1][0] = newInterval[0]
                    left = i + 1
                    break
                else:
                    intervals.insert(i+1, newInterval)
                    return intervals
        # if newInterval[1] <= intervals[left][1]:
        #     return intervals
        # else:
        for i in range(left, len(intervals)):
            if intervals[i][0] <= newInterval[1] <= intervals[i][1]:
                # right = i
                intervals[left:i+1] = [[intervals[left][0], intervals[i][1]]]
                return intervals
            if i < len(intervals) - 1 and intervals[i][1] < newInterval[1] < intervals[i+1][0]:
                intervals[left:i+1] = [[intervals[left][0], newInterval[1]]]
                return intervals
        intervals[left:] = [[intervals[left][0], newInterval[1]]]
        return intervals


if __name__ == "__main__":
    s = Solution()
    print()
